### ISSUE TITLE
forEach/map do not iterate over undefined values in IE 6-8

### ISSUE THREAD

**Issue by ljharb** [es-shims/es5-shim#190]
`var a = [undefined]; a.map(function (v) { return v; }).length === a.length` => true, which is good.

However, `var a = [undefined], b = []; a.forEach(function (v) { b.push(v); }); b.length === a.length` => false, which is bad (In Safari, for example, it gives true)

update: `0 in [undefined]` returns true for most browsers but false in IE 6-8, which is the root of the issue.

**Comment by mk-pmb** [comment]
@jareware, I got hit by this today, too. Turns out my time writing [my own shim collection](https://github.com/mk-pmb/epiphany-shims-pmb-js) wasn't wasted then. You can load that one after `es5-shim` and it will fix some of the array methods. :-)

**Comment by ljharb** [comment]
@mk-pmb a separate package doesn't help anyone; if you think you have a reliable way to distinguish `undefined` from `sparse` in IE 6 - 8, you'd be the first person to come up with one - and I'd love to accept a PR for that into es5-shim. (in other words, this is an IE 6 - 8 bug, not an es5-shim bug)

**Comment by mk-pmb** [comment]
> if you think you have a reliable way to distinguish undefined from sparse in IE 6 - 8,

Nope, I just wasn't precise really. My package just assumes arrays aren't sparse.

**Comment by ljharb** [comment]
Right - ES5 array methods do not assume arrays aren't sparse, because that would be a violation of the spec (just like assuming `undefined` is sparse is, but that's an engine bug that simply can't be worked around).

**Comment by mk-pmb** [comment]
At least it seems to be about array literals only. My MSIE distinguishes `undefined` from sparse at least when they're explicitly assigned:

```javascript
var a = [], i;
a[1] = undefined;
a[3] = undefined;
for (i = 0; i < 5; i += 1) { a[i] = (i in a); }
a.join(' ')  // in my MSIE 6: 'false true false true false'
```

**Comment by ljharb** [comment]
Indeed, you're right - `var a = [, undefined, , undefined, ]` prints out all falses.

I'll modify the sparse tests so that they can be run against those - but I'm pretty sure the es5-shim already does the correct thing in this instance.
